src/gin/nccl_ofi_gin.cpp

```diff
@@ -0,0 +1,734 @@
+
+	device->rdma_device_set_comm(comm_id, gin_comm);
+	gin_comm->rank_comms.resize(nranks);
+
+	ret = nccl_ofi_freelist_init_mr
+		(sizeof(nccl_net_ofi_rdma_signal_metadata_msg_t), 16, 16, 0, nullptr, nullptr,
+		 freelist_regmr_host_fn, freelist_deregmr_host_fn, rdma_ep->rdma_endpoint_get_domain(),
+		 1, &gin_comm->metadata_fl);
+	if (ret != 0) {
+		return ret;
+	}
+
+	ret = nccl_ofi_gin_allgather(gin_comm, all_handles.data(), sizeof(rdma_gin_connect_handle));
+	if (ret != 0) {
+		return ret;
+	}
+
+	for (int i = 0; i < nranks; ++i) {
+		const rdma_gin_connect_handle &gin_handle = all_handles[i];
+		nccl_ofi_gin_rank_comm &remote_rank_comm = gin_comm->rank_comms[i];
+		remote_rank_comm.comm_id = gin_handle.comm_id;
+		remote_rank_comm.next_target_seq_num = 0;
+		remote_rank_comm.flush_buff_addr = gin_handle.flush_buff_addr;
+
+		for (int r = 0; r < num_rails; ++r) {
+			ret = rail_addr_insert(rdma_ep->control_rails[r], gin_handle.control_ep_names[r], i,
+					       remote_rank_comm.control_address[r], gin_comm->ctrl_rank_map[r]);
+			if (ret != 0) {
+				return ret;
+			}
+
+			ret = rail_addr_insert(rdma_ep->rails[r], gin_handle.ep_names[r], i,
+					       remote_rank_comm.address[r], gin_comm->rank_map[r]);
+			if (ret != 0) {
+				return ret;
+			}
+			remote_rank_comm.flush_buff_mr_key[r] = gin_handle.flush_buff_mr_key[r];
+		}
+	}
+
+	(*gin_comm_out) = gin_comm;
+	return 0;
+}
+
+
+static inline int writedata_ack(nccl_ofi_gin_comm *gin_comm, unsigned int peer_rank,
+				unsigned int msg_seq_num)
+{
+	/* For now, always send acks on rail 0 */
+	const int rail_id = 0;
+
+	auto *req = new nccl_net_ofi_gin_writeack_req_t(gin_comm);
+
+	auto &rank_comm = gin_comm->rank_comms[peer_rank];
+	uint32_t peer_comm_id = rank_comm.comm_id;
+	uint32_t imm_data = GET_RDMA_WRITE_IMM_DATA(peer_comm_id, msg_seq_num, WRITEDATA_ACK_NSEG);
+
+	auto *rdma_ep = reinterpret_cast<nccl_net_ofi_rdma_ep_t *>(gin_comm->ep);
+	auto *domain = rdma_ep->rdma_endpoint_get_domain();
+
+	auto *ofi_ep = rdma_ep->rdma_endpoint_get_rail(rail_id)->ofi_ep.get();
+
+	auto *desc = fi_mr_desc(domain->flush_buff.mr_handle->mr[rail_id].get());
+
+	auto op = [=] {
+		ssize_t rc = fi_writedata(ofi_ep, domain->flush_buff.buffer, 0, desc, imm_data,
+					  rank_comm.address[0], rank_comm.flush_buff_addr,
+					  rank_comm.flush_buff_mr_key[rail_id], &(req->ctx.ofi_ctx));
+		if (rc != 0 && rc != -FI_EAGAIN) {
+			NCCL_OFI_WARN("Failed call to fi_writedata; RC: %zd", rc);
+		}
+		return rc;
+	};
+
+	int ret = op();
+	if (ret == -FI_EAGAIN) {
+		gin_comm->pending_requests.push_back(op);
+		ret = 0;
+	} else if (ret != 0) {
+		return ret;
+	}
+
+	gin_comm->outstanding_ack_counter++;
+
+	return ret;
+}
+
+
+static inline int do_gin_signal(nccl_ofi_gin_comm *gin_comm,
+				const nccl_net_ofi_rdma_signal_metadata_msg_t &metadata)
+{
+	void *signal_base = reinterpret_cast<void *>(metadata.signal_base_address);
+
+	uint64_t add_value = metadata.signal_value;
```

Comment:
Please add a comment to call out that the value is only relevant for add because the other supported op is increment.